external-crates/move/crates/move-compiler/src/unit_test/plan_builder.rs

```diff
@@ -543,26 +544,27 @@ fn parse_gas_budget_attribute(context: &mut Context, gas_budget_attr: Option<&E:
                     .add_diag(diag!(Attributes::InvalidValue, (*aloc, invalid_attr_msg)));
                 return None;
             }
-            let (gas_budget_kind, (attr_loc, attr)) = gas_budget_kind_vec.pop().unwrap();
-            match gas_budget_kind.as_str() {
-                TestingAttribute::GAS_BUDGET_COMPUTE_UNIT_LIMIT => {
-                    let (value_name_loc, attr_value) = get_assigned_attribute(
-                        context,
-                        TestingAttribute::GAS_BUDGET_COMPUTE_UNIT_LIMIT,
-                        attr_loc,
-                        attr,
-                    )?;
-                    let (_, _, u) =
-                        convert_constant_value_u64_constant_or_value(
-                            context,
-                            value_name_loc,
-                            &attr_value,
-                        )?;
-                    // TODO: Do some sanity check that u shouldn't be larger than a max value.
-                    assigned_gas_budget.compute_budget = u;
-                }
-                _ => unreachable!(),
-            };
+            while !gas_budget_kind_vec.is_empty() {
+                let (gas_budget_kind, (attr_loc, attr)) = gas_budget_kind_vec.pop().unwrap();
+                match gas_budget_kind.as_str() {
+                    TestingAttribute::GAS_BUDGET_COMPUTE_UNIT_LIMIT => {
+                        let u = get_assigned_attribute_as_u64(context, TestingAttribute::GAS_BUDGET_COMPUTE_UNIT_LIMIT, attr_loc, attr)?;
+                        // TODO: Do some sanity check that u shouldn't be larger than a max value.
+                        assigned_gas_budget.compute_budget = u;
+                    }
+                    TestingAttribute::GAS_BUDGET_HEAP_SIZE => {
+                        let u = get_assigned_attribute_as_u64(context, TestingAttribute::GAS_BUDGET_HEAP_SIZE, attr_loc, attr)?;
+                        // TODO: Do some sanity check that u shouldn't be larger than a max value.
+                        assigned_gas_budget.heap_size = u;
+                    }
+                    TestingAttribute::GAS_BUDGET_MAX_CALL_DEPTH => {
+                        let u = get_assigned_attribute_as_u64(context, TestingAttribute::GAS_BUDGET_MAX_CALL_DEPTH, attr_loc, attr)?;
+                        // TODO: Do some sanity check that u shouldn't be larger than a max value.
+                        assigned_gas_budget.max_call_depth = u;
+                    }
+                    _ => unreachable!(),
+                };
+            }
```

Comment:
Can it be this code instead?
```diff
             while !gas_budget_kind_vec.is_empty() {
                 let (gas_budget_kind, (attr_loc, attr)) = gas_budget_kind_vec.pop().unwrap();
                 match gas_budget_kind.as_str() {
-                    TestingAttribute::GAS_BUDGET_COMPUTE_UNIT_LIMIT => {
-                        let u = get_assigned_attribute_as_u64(context, TestingAttribute::GAS_BUDGET_COMPUTE_UNIT_LIMIT, attr_loc, attr)?;
-                        // TODO: Do some sanity check that u shouldn't be larger than a max value.
-                        assigned_gas_budget.compute_budget = u;
-                    }
-                    TestingAttribute::GAS_BUDGET_HEAP_SIZE => {
-                        let u = get_assigned_attribute_as_u64(context, TestingAttribute::GAS_BUDGET_HEAP_SIZE, attr_loc, attr)?;
+                    v if v == TestingAttribute::GAS_BUDGET_COMPUTE_UNIT_LIMIT ||
+                    v == TestingAttribute::GAS_BUDGET_HEAP_SIZE ||
+                    v == TestingAttribute::GAS_BUDGET_MAX_CALL_DEPTH => {
+                        let u = get_assigned_attribute_as_u64(context, v, attr_loc, attr)?;
                         // TODO: Do some sanity check that u shouldn't be larger than a max value.
                         assigned_gas_budget.heap_size = u;
-                    }
-                    TestingAttribute::GAS_BUDGET_MAX_CALL_DEPTH => {
-                        let u = get_assigned_attribute_as_u64(context, TestingAttribute::GAS_BUDGET_MAX_CALL_DEPTH, attr_loc, attr)?;
-                        // TODO: Do some sanity check that u shouldn't be larger than a max value.
-                        assigned_gas_budget.max_call_depth = u;

                     }
                     _ => unreachable!(),
                 };
```

Reply:
Also, if we match against str constants we should ensure that all constants are unique.
On simple eye inspection they are but in general I think it is not very safe.

Reply:
Ah, I did not spot the different assignment field.
Then can we do this?
```diff
             while !gas_budget_kind_vec.is_empty() {
                 let (gas_budget_kind, (attr_loc, attr)) = gas_budget_kind_vec.pop().unwrap();
-                match gas_budget_kind.as_str() {
+                let probe = gas_budget_kind.as_str();
+                match probe {
                     TestingAttribute::GAS_BUDGET_COMPUTE_UNIT_LIMIT => {
-                        let u = get_assigned_attribute_as_u64(context, TestingAttribute::GAS_BUDGET_COMPUTE_UNIT_LIMIT, attr_loc, attr)?;
+                        let u = get_assigned_attribute_as_u64(context, probe, attr_loc, attr)?;
                         // TODO: Do some sanity check that u shouldn't be larger than a max value.
                         assigned_gas_budget.compute_budget = u;
                     }
                     TestingAttribute::GAS_BUDGET_HEAP_SIZE => {
-                        let u = get_assigned_attribute_as_u64(context, TestingAttribute::GAS_BUDGET_HEAP_SIZE, attr_loc, attr)?;
+                        let u = get_assigned_attribute_as_u64(context, probe, attr_loc, attr)?;
                         // TODO: Do some sanity check that u shouldn't be larger than a max value.
                         assigned_gas_budget.heap_size = u;
                     }
                     TestingAttribute::GAS_BUDGET_MAX_CALL_DEPTH => {
-                        let u = get_assigned_attribute_as_u64(context, TestingAttribute::GAS_BUDGET_MAX_CALL_DEPTH, attr_loc, attr)?;
+                        let u = get_assigned_attribute_as_u64(context, probe, attr_loc, attr)?;
                         // TODO: Do some sanity check that u shouldn't be larger than a max value.
                         assigned_gas_budget.max_call_depth = u;
                     }
```

Reply:
`u` is an Option, but in each of matching cases the value is granted and we can simply unwrap.

Reply:
We match the same `gas_budget_kind.as_str()` as in the original code, and we have the same matching strings so the control flow does not change.

In all 3 cases `u` is Some, so it is safe to unwrap (since in the original code we unwrap).
Otherwise, whether it is Some or None we go to `unreachable` - same as it is now.